src/compiler/scala/tools/nsc/PhaseAssembly.scala

```diff
@@ -249,11 +249,12 @@ object DependencyGraph {
           graph.addEdge(after, p.phaseName, Follows)
         for (before <- p.runsBefore if before.nonEmpty && checkConstraint(before, "runsBefore"))
           graph.addEdge(p.phaseName, before, Follows)
+        def noPhase(s: String): Boolean = s.isEmpty || !graph.components.contains(s)
         if (p != start && p != end)
-          if (p.runsRightAfter.forall(_.isEmpty) && p.runsAfter.forall(_.isEmpty))
+          if (p.runsRightAfter.forall(noPhase) && p.runsAfter.forall(noPhase))
```

Comment:
I'm unsure this is the right solution. Such a lax contraint allows the phase to end up anywhere. The algorithm is deterministic, but the result depends on the phase name: `a-plugin` gives `parser - a-plugin - namer ...`, but `plugin` gives `parser - namer - packageobjects - typer - plugin ....`.

In the case of Scaladoc, this means that a plugin with `runsAfter pickler`, for example, could end up before typer and would likely crash.

2.13.14 drops phases where either the `runsAfter` or the `runsRightAfter` constraint is invalid (with a warning). The phase is also dropped when both `runsAfter` and `runsRightAfter` are empty (silently).

2.13.14 silently ignores invalid the `runsBefore` constraints. That's why the `"pickle"` typo in Scala.js didn't matter (https://github.com/scala-js/scala-js/commit/aa84f641c653fbbfc1b14378329cdf89cf1c5564).

So overall I suggest
  - drop phases with invalid `runsAfter`/`runsRightAfter` constraints, warn (also in scaladoc: if someone writes `runsAfter pickler` they should get notified that the phase is dropped in scaladoc. since we have that warning in 2.13.14 already, it's not a breaking change. maybe there should be a way for a phase to declare whether it's intended to also run in scaladoc, but that can be a follow-up.)
  - ignore invalid `runsBefore` constraints, warn unless in scaladoc. that is already done in this PR.

wdyt?

Reply:
> * drop phases with invalid `runsAfter`/`runsRightAfter` constraints, warn (also in scaladoc: if someone writes `runsAfter pickler` they should get notified that the phase is dropped in scaladoc. since we have that warning in 2.13.14 already, it's not a breaking change. maybe there should be a way for a phase to declare whether it's intended to also run in scaladoc, but that can be a follow-up.)

Won't that warn for any phase that declares otherwise valid constraints on later phases? Like
```scala
    override val runsAfter = List("specialize")
    override val runsBefore = List("explicitouter")
```
We have several phases like that in Scala.js, and I'm sure there are plenty of those in other plugins. Those should be dropped without warning. They are in 2.13.14.

Or actually give the plugins a way to declare that they should run in Scaladoc, and then only warn for those if they get dropped. Otherwise there will be lots of false positives.

Reply:
OK I've deleted the long reply I wrote while my oatmeal burned on the stove.

I'll do lrytz's suggestion with special dispensation for scalac phase names: IIUC, warn for unknown constraints, otherwise silently ignore phase that runsAfter a phase not in the run. (I'll check if a warning here is noisy.)

I'll check scala-js under 2.13.14 and head; I checked using Seth's instructions yesterday but did not check `-Vphases` or any output.

The principle is not to be "more helpful" but not to change anyone's perceived DX.

Reply:
Just to add, I liked the simplicity of this edit.

Reply:
> We have several phases like that in Scala.js, and I'm sure there are plenty of those in other plugins. Those should be dropped without warning. They are in 2.13.14.

OK - in my testing, 2.13.14 scaladoc warns on invalid `runsAfter` constraints:

```
➜ sandbox java -cp $(cs fetch -p org.scala-lang:scala-compiler:2.13.14):. scala.tools.nsc.ScalaDoc -usejavacp -Xplugin:plugin.jar T.scala -Vdebug
warning: dropping dependency on node with no phase object: specialize
[running phase parser on T.scala]
[running phase namer on T.scala]
[running phase packageobjects on T.scala]
[running phase typer on T.scala]
```

But you're probably right, the safest way forward is to never warn in Scaladoc.

Reply:
> lrytz's suggestion with special dispensation for scalac phase names

not sure which suggestion you mean; I like the current PR

Reply:
so we'd just have to drop phases with invalid `runsAfter` constraints, instead of putting `runsAfter = parser`

Reply:
https://fosstodon.org/@som_snytt/113080518511283717

I intended to confirm that running after a phantom phase is just dropped, but that is my current IIUC. My old comment would have been, No, constraints must be verified. If you intend a phase not to run, then don't configure it; or, the component can report that it won't be running because there is no phase X. But practically, it's OK to say "runs after" means "runs after X if X runs at all".

The lrytz suggestion I meant was to warn always ("there is no pickler in scaladoc") vs just for typos ("picklers").